swap/swap.go

```diff
@@ -256,13 +256,46 @@ func (s *Swap) createCheque(peer enode.ID) (*Cheque, error) {
 }
 
 //GetPeerBalance returns the balance for a given peer
-func (swap *Swap) GetPeerBalance(peer enode.ID) (int64, error) {
-	swap.lock.RLock()
-	defer swap.lock.RUnlock()
-	if p, ok := swap.balances[peer]; ok {
-		return p, nil
+func (s *Swap) GetPeerBalance(peer enode.ID) (int64, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	var peerBalance int64
+	var err error
+
+	// load balance from memory
+	peerBalance, keyExists := s.balances[peer]
+	if !keyExists {
+		// if not in memory, try to load balance from disk
+		err = s.stateStore.Get(peer.String(), &peerBalance)
+	}
+	return peerBalance, err
+}
+
+//GetAllBalances returns the balances for all known peers
+func (s *Swap) GetAllBalances() (map[enode.ID]int64, error) {
+	balances := make(map[enode.ID]int64)
+
+	// load balances from disk
+	peerIDs, err := s.stateStore.Keys()
```

Comment:
I assume that there will be a not so large number of peer IDs returned from the state store. Over time, if we do not cleanup "staled" peers, this list is only growing. What would be the accepted number of keys in memory? Would the iterator pattern or even sending over a channel make sense to avoid potential higher memory consumption? Or it is expected that even a high number of returned keys takes not significant amount of memory or makes no pressure to the gc if called frequently (every string here escapes to the heap, I suppose in this case).

Reply:
The number of entries will depend on how many peers we have a balance with. 

This means if we never limit the amount of peers we connect through the Kademlia, this list can become big. There was a `MaxPeer` constant at some point, do we still have it?

In normal operation the list should be manageable - up to 100 maybe?

However, if there is a lot of fluctuations in the network, specifically if there are attacks (peers who only suck out services up to payment threshold and then disconnect), this list can become pretty huge, and polluted with stale entries. In this case some archiving function or something, based on a timestamp, might be needed (check https://github.com/ethersphere/swarm/issues/1596).

In any case, the iteration would only be used when a user explicitly calls `GetAllBalances`, which is a manual operation and probably not too frequent.

Reply:
Thank you @janos for this very relevant observation.

I agree with @holisticode: let's deal with stale data in #1596.

Reply:
Thanks for the explanation @holisticode. As it is a manual operation, and not too frequent, it is not critical. And, I agree, @mortelli that it can be handled in some future iteration/pr. I just wanted to raise attention to it. Thanks.